storage/store.go

```diff
@@ -1407,40 +1431,91 @@ func (s *Store) resolveWriteIntentError(ctx context.Context, wiErr *proto.WriteI
 	}
 	wiErr.Resolved = true // success!
 
+	for i, intent := range pushIntents {
+		intent.Txn = *(bReply.Responses[i].InternalPushTxn.PusheeTxn)
+		resolveIntents = append(resolveIntents, intent)
+	}
+
+	rng.resolveIntents(ctx, resolveIntents)
+
+	return wiErr
+}
+
+// resolveIntents resolves the given intents. Those which are local to the range
+// use a fast-path by submitting directly to the range-local Raft instance; all
+// others are dispatched in a batch.
+func (r *Replica) resolveIntents(ctx context.Context, intents []proto.Intent) {
+	trace := tracer.FromCtx(ctx)
 	var wg sync.WaitGroup
 
 	// We pushed the transaction(s) successfully, so resolve the intent(s).
 	trace.Event("resolving intents [async]")
-	for i, intent := range wiErr.Intents {
-		pushReply := bReply.Responses[i].InternalPushTxn
-		intentKey := intent.Key
+	for i := range intents {
+		intent := intents[i] // avoids a race in `i, intent := range ...`
 		resolveArgs := &proto.InternalResolveIntentRequest{
 			RequestHeader: proto.RequestHeader{
 				// Use the pushee's timestamp, which might be lower than the
 				// pusher's request timestamp. No need to push the intent higher
 				// than the pushee's txn!
-				Timestamp: pushReply.PusheeTxn.Timestamp,
-				Key:       intentKey,
+				Timestamp: intent.Txn.Timestamp,
+				Key:       intent.Key,
 				User:      security.RootUser,
-				Txn:       pushReply.PusheeTxn,
+				Txn:       &intent.Txn,
 			},
 		}
 
+		// Either the intent is not local, in which case we add it to a batch
+		// request.
+		// TODO(tschottdorf): Lots to do here. This runs a batch for each
+		// individual external intent as preliminary code. Really we want a
+		// single batch (non-transactional) which contains the resolve requests
+		// for the individual intents. Those don't necessarily have the same
+		// transaction, so currently our batches don't support that - it's
+		// either a Txn on the batch and all requests share it, or nothing on
+		// the batch and nothing on any individual request. That can be
+		// changed, but there are some implications on tracing and on the
+		// transaction coordinator, so best left as a separate endeavor.
+		// TODO(tschottdorf): Also no support for key-range intents yet. We
+		// need to support those because the EndTransaction path will throw
+		// them our way. And we need a test that does exactly that.
+		// TODO(tschottdorf): currently synchronous. See if we still have that
```

Comment:
it's not important now. This is provisional code until the TODOs above are addressed. We'll want to do this async eventually.